internal/controllers/user_integrations_controller.go

```diff
@@ -58,14 +58,94 @@ func (udc *UserDevicesController) GetUserDeviceIntegration(c *fiber.Ctx) error {
 	apiIntegration, err := models.UserDeviceAPIIntegrations(
 		models.UserDeviceAPIIntegrationWhere.UserDeviceID.EQ(userDeviceID),
 		models.UserDeviceAPIIntegrationWhere.IntegrationID.EQ(integrationID),
+		qm.Load(models.UserDeviceAPIIntegrationRels.UserDevice),
 	).One(c.Context(), udc.DBS().Reader)
 	if err != nil {
 		if errors.Is(err, sql.ErrNoRows) {
 			return fiber.NewError(fiber.StatusBadRequest, fmt.Sprintf("user device %s does not have integration %s", userDeviceID, integrationID))
 		}
 		return err
 	}
-	return c.JSON(GetUserDeviceIntegrationResponse{Status: apiIntegration.Status, ExternalID: apiIntegration.ExternalID, CreatedAt: apiIntegration.CreatedAt})
+
+	resp := GetUserDeviceIntegrationResponse{
+		Status:     apiIntegration.Status,
+		ExternalID: apiIntegration.ExternalID,
+		CreatedAt:  apiIntegration.CreatedAt,
+	}
+	// Handle fetching virtual token status
+	intd, err := udc.DeviceDefSvc.GetIntegrationByID(c.Context(), integrationID)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid integration id")
+	}
+	if intd.Vendor == constants.TeslaVendor {
+		if !apiIntegration.ExternalID.Valid && apiIntegration.AccessToken.Valid && !(apiIntegration.R.UserDevice.VinConfirmed && apiIntegration.R.UserDevice.VinIdentifier.Valid) {
+			return fiber.NewError(fiber.StatusBadRequest, "missing device or integration details")
+		}
+		isConnected, err := udc.getDeviceVirtualTokenStatus(c.Context(), apiIntegration)
+		if err != nil {
+			return fiber.NewError(fiber.StatusInternalServerError, fmt.Sprintf("error checking verifying tesla connection status %s", err.Error()))
+		}
+
+		resp.Tesla = TeslaConnectionStatus{
+			IsVirtualTokenConnected: isConnected,
+		}
+	}
+
+	return c.JSON(resp)
+}
+
+func (udc *UserDevicesController) getDeviceVirtualTokenStatus(ctx context.Context, integration *models.UserDeviceAPIIntegration) (bool, error) {
+	accessTk := integration.AccessToken.String
+
+	meta := services.UserDeviceAPIIntegrationsMetadata{}
+	err := integration.Metadata.Unmarshal(&meta)
+	if err != nil {
+		return false, errors.New("error occurred fetching virtual token status")
+	}
+
+	if meta.TeslaRegion == "" {
+		return false, errors.New("missing tesla region")
+	}
+
+	isNewToken, auth, err := udc.ensureTeslaTokenValidty(ctx, integration.RefreshToken.String, integration.AccessExpiresAt.Time)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get tesla credentials")
+	}
+
+	if isNewToken {
```

Comment:
Should we update the KTable, too, if we do this? I'm also a little worried about races.

Is the plan to shut off the pollers if telemetry streaming is enabled?

Reply:
We will not be shutting off the pollers, they will be complementing this actually. The Telemetry server is limited in what it allows us listen for. Also I don't think we need to update the KTables since we won't be using this anywhere else or calling it frequently, this is just to allow the frontend discover what can be activated for an integration type, the next [ticket](https://linear.app/dimo/issue/SI-2558/create-endpoint-to-enable-telemetry-for-an-integration), handles registering a vehicle for telemetry based on what the integration is capable of.

Reply:
We talked a bit—I think we agree that the KTable could end up with a refresh token that is unusable. More to come.